### PR TITLE
feat(ci): add Vercel prebuilt deployment primitives

### LINKED ISSUE #516: ci(vercel): add tested deployment planning and prebuilt-build primitives
Parent: #515

## Outcome

Add deterministic, unit-tested primitives that answer two questions without contacting Vercel:

1. Which of the four Vercel applications are affected by a Git diff?
2. What stable, collision-resistant deployment ID should a prebuilt Next.js build use?

This issue must not deploy anything or change which system owns deployments.

## Why this is separate

The deployment workflows will handle credentials and production domains. The planning layer should remain testable with ordinary fixtures so future changes do not accidentally skip a required preview or deploy all four applications for documentation-only changes.

## Current application graph

| Deployable | Workspace | Direct shared-package dependency |
|---|---|---|
| app.mento.org | `apps/app.mento.org` | `packages/ui`, `packages/web3` |
| governance.mento.org | `apps/governance.mento.org` | `packages/ui`, `packages/web3` |
| reserve.mento.org | `apps/reserve.mento.org` | `packages/ui`, `packages/web3` |
| ui.mento.org | `apps/ui.mento.org` | `packages/ui` |

Do not hard-code only this table and assume it can never drift. Use Turborepo's package graph for normal package changes, plus a small reviewed fail-closed override list for root/global inputs whose effect Turbo cannot safely infer.

### 1. Pin the Vercel CLI

Use an exact version, not a range. The latest stable observed while scoping on 2026-07-14 was `56.2.0`; re-query the published stable version when implementing and record the chosen version in the PR.

The repository policy requires explicit owner approval before introducing a new npm dependency. Obtain that approval before adding `vercel` as a root devDependency. Do not silently use `npx vercel@latest`.

### 2. Add an affected-deployment planner

Add a small script and tests. Choose an unambiguous name such as:

- `scripts/plan-vercel-deployments.mjs`
- `scripts/plan-vercel-deployments.test.mjs`

Inputs must be explicit and reproducible:

- base commit SHA;
- head commit SHA;
- a repository with enough history to resolve both commits.

Use Turborepo's SCM-aware affected mode (`TURBO_SCM_BASE`, `TURBO_SCM_HEAD`, and `turbo run build --affected --dry=json`) or an equivalently tested package-graph query. Do not derive the base from “whatever happens to be checked out.”

Keep the planner entry point self-contained enough to execute from a trusted base-SHA copy in PR workflows, or document the complete helper-module set that must be copied from that SHA. It must never import classifier logic from the untrusted PR head.

Output a stable machine-readable object containing an ordered `deployments` array. The only valid deployment names are:

- `app`
- `governance`
- `reserve`
- `ui`

Example:

```json
{"deployments":["app","reserve"],"base":"<sha>","head":"<sha>","reason":"affected-packages"}
```

### 3. Fail closed for ambiguous/global changes

Return all four deployments when:

- either SHA is missing or cannot be resolved;
- Turbo planning fails or returns malformed output;
- the workflow/deployment planner itself changes;
- a root build input with cross-workspace impact changes (for example `pnpm-lock.yaml`, `turbo.json`, root package manager configuration, patches, or security-header/build configuration shared by the apps);
- the planner cannot confidently classify the change.

Return no deployments for proven docs-only/test-only changes that cannot affect a deployed artifact. Encode those paths in tests; do not implement a broad “unknown files are ignored” rule.

Expected graph cases:

- an app-local build change deploys that app only;
- `packages/ui` deploys all four;
- `packages/web3` deploys app, governance, and reserve;
- a root/global build input deploys all four;
- docs-only changes deploy none;
- planner uncertainty deploys all four.

### 4. Add a deployment-ID helper

Prebuilt Next.js deployments need a unique deployment ID for Skew Protection. Vercel rejects reuse of an ID within the same project, including a rerun of the same SHA. Add a tested helper that derives the ID from the target, immutable commit SHA, and immutable workflow execution identity (at minimum GitHub `run_id` plus `run_attempt`). Generate it once per workflow attempt. The ID belongs to the prebuilt output: `vercel deploy` has no deployment-ID flag. Expose the generated value to each app's `next.config.ts` during `vercel build`, assert that `.vercel/output/config.json` contains exactly that value, and upload that unchanged `.vercel/output` with `vercel deploy --prebuilt`. Never regenerate the ID between build and upload or invent a deploy-CLI flag. It must:

- be at most 32 characters;
- contain only `A-Z`, `a-z`, `0-9`, `_`, and `-`;
- not begin with Vercel's reserved `dpl_` prefix;
- differ across deployable targets in the same workflow run;
- differ when the same target/SHA is rerun or retried;
- be identical when recomputed for the same target, SHA, run ID, and run attempt.

Wire one explicitly named deployment-ID environment input into each Next.js configuration and include it in the relevant Turbo build inputs so a cached build cannot accidentally reuse another deployment's ID. Add a test/assertion for the resulting `.vercel/output/config.json`; deployment code consumes the unchanged prebuilt directory and accepts no second deployment-ID input.

As part of implementation, assert and record that the resolved Next.js version is newer than `16.2.0-canary.15` and the pinned Vercel CLI is newer than `50.3.3`, the current documented prerequisites for custom Next.js deployment IDs. The repository currently declares Next `^16.2.10`, but verify the lockfile-resolved version rather than relying on this ticket's snapshot.

### 5. Audit build-time environment semantics

Document every build-time read of:

- `VERCEL_ENV`
- `VERCEL_TARGET_ENV`
- `NEXT_PUBLIC_VERCEL_ENV`
- other Vercel system variables required by these apps

Extend the inventory to every required application build variable. At minimum, cover `SENTRY_AUTH_TOKEN` for production-semantics builds of app/governance/reserve and governance's required `ETHERSCAN_API_KEY`. For every target and environment, classify each variable as system-injected, pullable by `vercel pull`, a safe explicit constant, or Vercel Sensitive/non-exportable. A Sensitive value is non-readable after creation, so the workflow must not assume `vercel pull` can recover it. Any required non-exportable value must have a maintainer-managed GitHub secret mapping in #517 at the narrowest preview/production scope. Missing-variable checks must fail before the build and print only the variable name, never its value.

The later workflow must explicitly restore values that Vercel normally injects and any required non-exportable build variables, because `vercel deploy --prebuilt` does not provide system environment variables during the local build.

Preserve the current app behavior: `app.mento.org` main is built for the custom `v3` target, where `VERCEL_ENV=preview` and `VERCEL_TARGET_ENV=v3`. Changing Sentry source-map behavior for that target is a separate product/observability decision.

### 6. Document the primitives

Add a deployment runbook (for example `docs/vercel-deployments.md`) and update `CLAUDE.md`/other command documentation required by the repository's docs-drift rule.

## Tests

Add fixture-driven tests that run without network access and cover, at minimum:

- each app-local change;
- `packages/ui`;
- `packages/web3`;
- root lockfile/global configuration;
- docs-only;
- invalid/missing base;
- invalid planner output;
- stable output ordering;
- every deployment-ID constraint above, including same-SHA rerun uniqueness and within-attempt stability;
- generated-ID propagation into `.vercel/output/config.json` and rejection of a mismatched/missing output ID;
- the per-target/environment build-variable classification, including a Sensitive/non-exportable fixture and a missing-variable failure that does not reveal values.

## Verification commands

The implementation PR must list and run the exact commands it adds. At minimum:

```bash
pnpm install --frozen-lockfile
pnpm test
pnpm build
```

If the new tests have a narrower command, run that first and document it. Also run the repository's normal formatting/lint/typecheck gates affected by the files changed.

## Acceptance criteria

- [ ] An exact Vercel CLI version is recorded after owner approval.
- [ ] Planning uses explicit immutable base/head SHAs and sufficient Git history.
- [ ] The trusted base-SHA planner can execute without importing classifier code from the PR head and emits stable JSON with only the four allowed target names.
- [ ] Ambiguous/global changes fail closed to all four targets.
- [ ] Proven non-runtime changes can yield an empty plan.
- [ ] Package graph cases above are covered by deterministic tests.
- [ ] Unique Next deployment IDs meet every documented constraint, differ on same-SHA reruns, remain stable within one build/upload attempt, and affect the build hash.
- [ ] The generated ID is present in `.vercel/output/config.json`; `vercel deploy --prebuilt` uploads that unchanged output and receives no invented ID flag/input.
- [ ] Resolved Next.js and pinned Vercel CLI versions satisfy Vercel's custom-deployment-ID prerequisites.
- [ ] Vercel build-time environment semantics, including app `v3`, are documented.
- [ ] Every required build-time application variable is classified by target/environment, and every non-exportable value has an explicit GitHub secret mapping prerequisite in #517.
- [ ] No Vercel API call, deployment, alias, or project ownership change is made.
- [ ] New commands/workflows are reflected in all relevant repo docs.

## References

- [Turborepo `--affected`](https://turborepo.dev/docs/reference/run#--affected)
- [Vercel CLI `build`](https://vercel.com/docs/cli/build)
- [Vercel CLI `deploy --prebuilt`](https://vercel.com/docs/cli/deploy)
- [Vercel system environment variables](https://vercel.com/docs/environment-variables/system-environment-variables)
- [Vercel Skew Protection](https://vercel.com/docs/skew-protection)

### PR DESCRIPTION
**Pull request by chapati23**
## The Problem

- Moving Vercel compilation to GitHub Actions needs trusted, deterministic primitives before any workflow receives deployment credentials. Without them, path-aware previews could silently skip an affected app, and prebuilt Next.js outputs could reuse an invalid deployment identity.

## The Solution

- Add a fail-closed affected-deployment planner backed by Turborepo's real package graph, with narrow reviewed exemptions only for proven non-runtime paths.
- Add deterministic per-target/per-attempt Next.js deployment IDs and verify the exact ID embedded in `.vercel/output/config.json`.
- Pin Vercel CLI `56.2.0`, verify resolved Next.js/Vercel prerequisites, inventory required build variables and Vercel environment semantics, and document the canonical migration runbook.
- Constrain transitive Vercel CLI dependencies to patched releases and narrowly suppress two OSV false positives that refer to an unrelated legacy package with the same `sandbox` name.
- Wire the deployment-ID input through all four Next.js configs and Turbo cache inputs without changing deployment ownership or contacting Vercel.

## Validation

- `CI=true pnpm install --frozen-lockfile`
- `pnpm vercel:primitives:test` — 50/50 passed, including real installed-Turbo graph fixtures, app-root Vercel environment loading, and adversarial runtime-looking path cases
- `pnpm vercel:versions:check` — Next.js `16.2.10`, Vercel CLI `56.2.0`
- `osv-scanner scan --config osv-scanner.toml --lockfile pnpm-lock.yaml --format table` — no findings and no unused ignores
- `pnpm test` — 6/6 Turbo tasks passed
- `pnpm check-types` — 8/8 Turbo tasks passed when run serially
- `pnpm ci:action-pins` and `pnpm ci:action-pins:test` — passed
- `pnpm quality:budgets:test` — 30/30 passed
- `trunk fmt`, `trunk check`, and `git diff --check` — passed
- Autoreview found and fixed the broad test/Markdown skip rule, mocked graph coverage, prerelease comparison, and linked-project env-file lookup; the deterministic follow-up review is clean.
- `pnpm build` — 6/6 Turbo tasks passed outside the filesystem sandbox (Turbopack requires an internal loopback listener).
- A real custom-ID Next.js build verified the routes manifest/config ID and conditional `runtimeServerDeploymentId: false` workaround.
- No user-facing UI behavior changed. Chrome DevTools MCP is unavailable in this session, so there was no applicable browser interaction to verify.

## Ship Checklist

- [x] PR title follows the [conventions](https://www.notion.so/Git-Branching-and-Commit-Message-Conventions-18f66f7d06444cfcbac5725ffbc7c04a?pvs=4#9355048863c549ef92fe210a8a1298aa)
- [x] Performed a self-review of my own changes
- [x] Relevant automated checks and smoke tests pass

Closes #516

### CODE REVIEW

File: apps/app.mento.org/next.config.ts [hunk 1]
```diff
@@ -74,6 +77,7 @@ const reportOnlyCsp = [
 ].join("; ");
 
 const nextConfig: NextConfig = {
+  deploymentId,
```
[the remark this answers is not in the document]
Comment (chapati23) [reply apps/app.mento.org/next.config.ts:80]: Fixed in f6711d26 — all four apps conditionally set experimental.runtimeServerDeploymentId to false when the custom prebuilt deployment ID is present, while native Vercel Git builds remain unchanged. A real Next.js build verified both artifact values.

File: scripts/vercel-build-environment.mjs [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    sensitive("SENTRY_AUTH_TOKEN", {
+      environments: ["production"],
+      githubScope: "vercel-cli-production environment; app build step only",
+    }),
+  ],
+  governance: [
+    pullable("NEXT_PUBLIC_BLOCKSCOUT_API_URL"),
+    pullable("NEXT_PUBLIC_BLOCKSCOUT_GRAPHQL_URL"),
+    pullable("NEXT_PUBLIC_ETHERSCAN_API_URL"),
+    pullable("NEXT_PUBLIC_GRAPH_API_KEY", { allowEmpty: true }),
```
[the remark this answers is not in the document]
Comment (chapati23) [reply scripts/vercel-build-environment.mjs]: Fixed in f6711d26 — NEXT_PUBLIC_GRAPH_API_KEY now rejects empty values, with a regression test for the governance preview contract.

File: scripts/vercel-prebuilt.mjs [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+  }
+
+  if (!Object.hasOwn(config, "deploymentId")) {
+    throw new Error("Prebuilt config is missing deploymentId");
+  }
+  if (config.deploymentId !== expectedId) {
+    throw new Error("Prebuilt deploymentId does not match the generated ID");
+  }
+  return expectedId;
+}
```
[the remark this answers is not in the document]
Comment (chapati23) [reply scripts/vercel-prebuilt.mjs:80]: Won't fix: pinned Vercel CLI 56.2.0 reads the Next.js deploymentId from .next/routes-manifest.json and writes it into the final .vercel/output/config.json; the official Build Output API contract also defines deploymentId there. The assertion intentionally verifies the exact config uploaded by --prebuilt. Clarified this propagation in f6711d26.

File: scripts/vercel-build-environment.mjs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    target,
+    environment,
+  }));
+}
+
+export function validateVercelBuildEnvironment({
+  target,
+  environment,
+  values = process.env,
+}) {
```
[the remark this answers is not in the document]
Comment (chapati23) [reply scripts/vercel-build-environment.mjs:138]: Fixed in f6711d26 — the checker now requires the exact Vercel project directory, loads its .vercel/.env.<environment>.local file, and overlays explicit workflow constants and scoped GitHub secrets. A root-CLI app-directory fixture covers the real monorepo layout.

File: scripts/vercel-build-environment.mjs [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    }),
+    sensitive("SENTRY_AUTH_TOKEN", {
+      environments: ["production"],
+      githubScope:
+        "vercel-cli-production environment; governance build step only",
+    }),
+  ],
+  reserve: [
+    pullable("NEXT_PUBLIC_STORAGE_URL"),
+    pullable("NEXT_PUBLIC_SENTRY_DSN_RESERVE", { allowEmpty: true }),
```
[the remark this answers is not in the document]
Comment (chapati23) [reply scripts/vercel-build-environment.mjs:74]: Fixed in f6711d26 — NEXT_PUBLIC_ANALYTICS_API_URL is now a required non-empty Reserve pull variable, with regression coverage and corrected runbook inventory.

**Comment by chapati23** [comment]
Review sweep complete for f6711d26. All five inline findings were triaged and replied to: four were fixed with regression coverage, and the config.json finding was declined with pinned Vercel CLI 56.2.0 plus official Build Output API evidence. All addressed threads are resolved; 50/50 primitive tests, full tests, typechecking, real custom-ID Next build, supply-chain checks, preview smoke, visual regression, and connected E2E are green. Fresh semantic review found no remaining issues.
